examples/rcn.py

```diff
@@ -0,0 +1,452 @@
+
+import jax
+import matplotlib.pyplot as plt
+import numpy as np
+from jax import numpy as jnp
+from jax import tree_util
+from joblib import Memory
+
+memory = Memory("./example_data/tmp")
+
+from scipy.ndimage import maximum_filter
+from scipy.signal import fftconvolve
+from sklearn.datasets import fetch_openml
+
+from pgmax.fg import graph, groups
+
+fetch_openml_cached = memory.cache(fetch_openml)
+os.environ["XLA_PYTHON_CLIENT_ALLOCATOR"] = "platform"
+os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "false"
+
+# %% [markdown]
+# # 1. Load the data
+# %% [markdown]
+# In this notebook, we train an RCN model on a small subset of MNIST (training on 20 images and testing on 20 images). We also have support to train the model on 100 images. Use this option only if you have a gpu with atleast 8Gbs memory. Also, recommend that jax is installed with cuda enabled for training with more than 20 images.
+
+# %%
+train_size = 20
+test_size = 20
+
+
+# %%
+def fetch_mnist_dataset(test_size: int, seed: int = 5) -> Tuple[np.ndarray, np.ndarray]:
+    """Returns test images sampled randomly from the set of MNIST images.
+
+    Args:
+        test_size: Desired number of test images.
+        seed: An integer used for repeatable tests.
+
+    Returns:
+        test_set: A list of length test_size containing images from the MNIST test dataset.
+        test_labels: Corresponding labels for the test images.
+    """
+
+    mnist_train_size = 60000
+    num_per_class = test_size // 10
+
+    print("Fetching the MNIST dataset")
+    dataset = fetch_openml_cached("mnist_784", as_frame=False, cache=True)
+    print("Successfully downloaded the MNIST dataset")
+
+    mnist_images = dataset["data"]
+    mnist_labels = dataset["target"].astype("int")
+    full_mnist_test_images = mnist_images[mnist_train_size:]
+    full_mnist_test_labels = mnist_labels[mnist_train_size:]
+
+    np.random.seed(seed)
+
+    test_set = []
+    test_labels = []
+    for i in range(10):
+        idxs = np.random.choice(
+            np.argwhere(full_mnist_test_labels == i)[:, 0], num_per_class
+        )
+        for idx in idxs:
+            img = full_mnist_test_images[idx].reshape(28, 28)
+            img_arr = jax.image.resize(image=img, shape=(112, 112), method="bicubic")
+            img = jnp.pad(
+                img_arr,
+                pad_width=tuple([(p, p) for p in (44, 44)]),
+                mode="constant",
+                constant_values=0,
+            )
+
+            test_set.append(img)
+            test_labels.append(i)
+
+    return np.array(test_set), np.array(test_labels)
+
+
+# %%
+test_set, test_labels = fetch_mnist_dataset(test_size)
+
+# %% [markdown]
+# # 2. Load the model
+
+# %% [markdown]
+# We load the following variables (that have been pre-computed).
+# - train_set and train_labels - A sample of MNIST train dataset containing 100 train images and their labels.
+# - frcs and edges - Used to represent the learned rcn graphical models.
```

Comment:
Link to the learning module in science RCN repo?

Reply:
Added.